rio/src/RIO/File.hs

```diff
@@ -0,0 +1,428 @@
+fsyncFileDescriptor name cFd =
+  liftIO $
+  void $
+    throwErrnoIfMinus1 ("fsync - " <> name) $
+    c_safe_fsync cFd
+
+-- NOTE: Not Async exception safe
+openFileFromDir :: (MonadIO m) => Fd -> FilePath -> IOMode -> m Handle
+openFileFromDir (Fd dirFd) fp iomode =
+  liftIO $
+  withFilePath fp $ \f -> do
+    bracketOnError
+      (do fileFd <- throwErrnoIfMinus1Retry "openFileFromDir" $
+                      c_safe_openat dirFd f (ioModeToFlags iomode) 0o666
+          FD.mkFD
+             fileFd
+             iomode
+             Nothing {- no stat -}
+             False {- not a socket -}
+             False {- non_blocking -}
+            `onException` c_close fileFd)
+      (liftIO . Device.close . fst)
+      (\(fD, fd_type) -> do
+         -- we want to truncate() if this is an open in WriteMode, but only if the
+         -- target is a RegularFile. ftruncate() fails on special files like
+         -- /dev/null.
+         when (iomode == WriteMode && fd_type == RegularFile) $
+           Device.setSize fD 0
+         HandleFD.mkHandleFromFD fD fd_type fp iomode False Nothing)
+
+-- | Opens a file using the openat C low-level API. This approach allows us to
+-- get a file descriptor for the directory that contains the file, which we can
+-- use later on to fsync the directory with.
+--
+-- @since 0.1.6
+openFileAndDirectory :: MonadUnliftIO m => FilePath -> IOMode -> m (Fd, Handle)
+openFileAndDirectory absFp iomode =  do
+  let dir = takeDirectory absFp
+      fp = takeFileName absFp
+
+  dirFd <- openDir dir
+  fileHandle <- openFileFromDir dirFd fp iomode `onException` closeDirectory dirFd
+  return (dirFd, fileHandle)
+
+-- | This sub-routine does the following tasks:
+--
+-- * It calls fsync and then closes the given Handle (mapping to a temporal/backup filepath)
+-- * It calls fsync and then closes the containing directory of the file
+--
+-- These steps guarante that the file changes are durable.
+--
+-- @since 0.1.6
+closeFileDurable :: MonadIO m => Fd -> Handle -> m ()
+closeFileDurable dirFd@(Fd cDirFd) h =
+  liftIO $
+  finally
+    (do fileFd <- handleToFd h
+        fsyncFileDescriptor "closeFileDurable/File" (FD.fdFD fileFd) `finally`
+          hClose h
+        -- NOTE: Here we are purposefully not fsyncing the directory if the file fails to fsync
+        fsyncFileDescriptor "closeFileDurable/Directory" cDirFd)
+    (closeDirectory dirFd)
+
+buildTemporaryFilePath :: MonadUnliftIO m => FilePath -> m FilePath
+buildTemporaryFilePath filePath = do
+  let
+    dirFp  = takeDirectory filePath
+    fileFp = takeFileName filePath
+  bracket (liftIO $ openBinaryTempFile dirFp fileFp)
+          (hClose . snd)
+          (return . fst)
+
+toTmpFilePath :: MonadUnliftIO m => FilePath -> m FilePath
+toTmpFilePath filePath =
+    buildTemporaryFilePath (dirPath </> tmpFilename)
+  where
+    dirPath = takeDirectory filePath
+    filename = takeFileName filePath
+    tmpFilename = "." <> filename <> ".tmp"
```

Comment:
I thought this was good giving I'm relying on the `openBinaryTempFile` call on line 262 to ensure uniqueness. 

I took a look [at its implementation](http://hackage.haskell.org/package/base-4.12.0.0/docs/src/System.IO.html#openTempFile'), and I'm noticing it will loop until it finds a unique name. Even in the situation of a race condition, the original algorithm generates a random String for each attempt (it keeps a global IORef counter to change a suffix number on each call).